File: pmb/ai_manage/models/held_product.py
"""用户持有产品数据模型"""
from dataclasses import dataclass, field


@dataclass
class HeldWealthProduct:
    """持有理财产品"""
    id: str
    holder_name: str
    product_name: str
    category: str
    term: str
    card_tail: str
    holding_amount: float
    holding_profit: float
    principal: float
    shares: float
    yield_rate: float
    redeem_date: str
    redeem_amount: float
    redeem_status: str

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "holder_name": self.holder_name,
            "product_name": self.product_name,
            "category": self.category,
            "term": self.term,
            "card_tail": self.card_tail,
            "holding_amount": self.holding_amount,
            "holding_profit": self.holding_profit,
            "principal": self.principal,
            "shares": self.shares,
            "yield_rate": self.yield_rate,
            "redeem_date": self.redeem_date,
            "redeem_amount": self.redeem_amount,
            "redeem_status": self.redeem_status,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "HeldWealthProduct":
        return cls(
            id=str(d.get("id", "")),
            holder_name=str(d.get("holder_name", "")),
            product_name=str(d.get("product_name", "")),
            category=str(d.get("category", "")),
            term=str(d.get("term", "")),
            card_tail=str(d.get("card_tail", "")),
            holding_amount=float(d.get("holding_amount", 0) or 0),
            holding_profit=float(d.get("holding_profit", 0) or 0),
            principal=float(d.get("principal", 0) or 0),
            shares=float(d.get("shares", 0) or 0),
            yield_rate=float(d.get("yield_rate", 0) or 0),
            redeem_date=str(d.get("redeem_date", "")),
            redeem_amount=float(d.get("redeem_amount", 0) or 0),
            redeem_status=str(d.get("redeem_status", "")),
        )


@dataclass
class HeldLoan:
    """持有贷款"""
    id: str
    holder_name: str
    loan_type: str
    loan_no: str
    purpose: str
    bank_branch: str
    issue_date: str
    principal: float
    remaining_principal: float
    repaid_principal: float
    rate_type: str
    annual_rate: str
    lpr_adjust: str
    next_repricing_date: str
    repayment_method: str
    repayment_card: str
    notify_phone: str

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "holder_name": self.holder_name,
            "loan_type": self.loan_type,
            "loan_no": self.loan_no,
            "purpose": self.purpose,
            "bank_branch": self.bank_branch,
            "issue_date": self.issue_date,
            "principal": self.principal,
            "remaining_principal": self.remaining_principal,
            "repaid_principal": self.repaid_principal,
            "rate_type": self.rate_type,
            "annual_rate": self.annual_rate,
            "lpr_adjust": self.lpr_adjust,
            "next_repricing_date": self.next_repricing_date,
            "repayment_method": self.repayment_method,
            "repayment_card": self.repayment_card,
            "notify_phone": self.notify_phone,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "HeldLoan":
        return cls(
            id=str(d.get("id", "")),
            holder_name=str(d.get("holder_name", "")),
            loan_type=str(d.get("loan_type", "")),
            loan_no=str(d.get("loan_no", "")),
            purpose=str(d.get("purpose", "")),
            bank_branch=str(d.get("bank_branch", "")),
            issue_date=str(d.get("issue_date", "")),
            principal=float(d.get("principal", 0) or 0),
            remaining_principal=float(d.get("remaining_principal", 0) or 0),
            repaid_principal=float(d.get("repaid_principal", 0) or 0),
            rate_type=str(d.get("rate_type", "")),
            annual_rate=str(d.get("annual_rate", "")),
            lpr_adjust=str(d.get("lpr_adjust", "")),
            next_repricing_date=str(d.get("next_repricing_date", "")),
            repayment_method=str(d.get("repayment_method", "")),
            repayment_card=str(d.get("repayment_card", "")),
            notify_phone=str(d.get("notify_phone", "")),
        )


@dataclass
class HeldPension:
    """养老金账户"""
    id: str
    holder_name: str
    account_type: str
    total_amount: float
    idle_amount: float
    annual_deposit_limit: float
    remaining_deposit_quota: float
    deposited_amount: float
    tax_benefit: str
    auto_deposit_status: str
    recommended_plan: str
    reference_yield: str
    yield_calculation: str
    remark: str

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "holder_name": self.holder_name,
            "account_type": self.account_type,
            "total_amount": self.total_amount,
            "idle_amount": self.idle_amount,
            "annual_deposit_limit": self.annual_deposit_limit,
            "remaining_deposit_quota": self.remaining_deposit_quota,
            "deposited_amount": self.deposited_amount,
            "tax_benefit": self.tax_benefit,
            "auto_deposit_status": self.auto_deposit_status,
            "recommended_plan": self.recommended_plan,
            "reference_yield": self.reference_yield,
            "yield_calculation": self.yield_calculation,
            "remark": self.remark,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "HeldPension":
        return cls(
            id=str(d.get("id", "")),
            holder_name=str(d.get("holder_name", "")),
            account_type=str(d.get("account_type", "")),
            total_amount=float(d.get("total_amount", 0) or 0),
            idle_amount=float(d.get("idle_amount", 0) or 0),
            annual_deposit_limit=float(d.get("annual_deposit_limit", 0) or 0),
            remaining_deposit_quota=float(d.get("remaining_deposit_quota", 0) or 0),
            deposited_amount=float(d.get("deposited_amount", 0) or 0),
            tax_benefit=str(d.get("tax_benefit", "")),
            auto_deposit_status=str(d.get("auto_deposit_status", "")),
            recommended_plan=str(d.get("recommended_plan", "")),
            reference_yield=str(d.get("reference_yield", "")),
            yield_calculation=str(d.get("yield_calculation", "")),
            remark=str(d.get("remark", "")),
        )

File: aipmb_app/lib/models/held_product.dart
/// 持有理财产品
class HeldWealthProduct {
  final String id;
  final String productName;
  final String category;
  final String term;
  final String cardTail;
  final double holdingAmount;
  final double holdingProfit;
  final double principal;
  final double shares;
  final double yieldRate;
  final String redeemDate;
  final double redeemAmount;
  final String redeemStatus;

  HeldWealthProduct({
    required this.id,
    required this.productName,
    required this.category,
    required this.term,
    required this.cardTail,
    required this.holdingAmount,
    required this.holdingProfit,
    required this.principal,
    required this.shares,
    required this.yieldRate,
    required this.redeemDate,
    required this.redeemAmount,
    required this.redeemStatus,
  });

  factory HeldWealthProduct.fromJson(Map<String, dynamic> json) => HeldWealthProduct(
    id: json['id']?.toString() ?? '',
    productName: json['product_name']?.toString() ?? '',
    category: json['category']?.toString() ?? '',
    term: json['term']?.toString() ?? '',
    cardTail: json['card_tail']?.toString() ?? '',
    holdingAmount: _parseDouble(json['holding_amount']),
    holdingProfit: _parseDouble(json['holding_profit']),
    principal: _parseDouble(json['principal']),
    shares: _parseDouble(json['shares']),
    yieldRate: _parseDouble(json['yield_rate']),
    redeemDate: json['redeem_date']?.toString() ?? '',
    redeemAmount: _parseDouble(json['redeem_amount']),
    redeemStatus: json['redeem_status']?.toString() ?? '',
  );
}

/// 持有贷款
class HeldLoan {
  final String id;
  final String loanType;
  final String loanNo;
  final String purpose;
  final String bankBranch;
  final String issueDate;
  final double principal;
  final double remainingPrincipal;
  final double repaidPrincipal;
  final String rateType;
  final String annualRate;
  final String lprAdjust;
  final String nextRepricingDate;
  final String repaymentMethod;
  final String repaymentCard;
  final String notifyPhone;

  HeldLoan({
    required this.id,
    required this.loanType,
    required this.loanNo,
    required this.purpose,
    required this.bankBranch,
    required this.issueDate,
    required this.principal,
    required this.remainingPrincipal,
    required this.repaidPrincipal,
    required this.rateType,
    required this.annualRate,
    required this.lprAdjust,
    required this.nextRepricingDate,
    required this.repaymentMethod,
    required this.repaymentCard,
    required this.notifyPhone,
  });

  factory HeldLoan.fromJson(Map<String, dynamic> json) => HeldLoan(
    id: json['id']?.toString() ?? '',
    loanType: json['loan_type']?.toString() ?? '',
    loanNo: json['loan_no']?.toString() ?? '',
    purpose: json['purpose']?.toString() ?? '',
    bankBranch: json['bank_branch']?.toString() ?? '',
    issueDate: json['issue_date']?.toString() ?? '',
    principal: _parseDouble(json['principal']),
    remainingPrincipal: _parseDouble(json['remaining_principal']),
    repaidPrincipal: _parseDouble(json['repaid_principal']),
    rateType: json['rate_type']?.toString() ?? '',
    annualRate: json['annual_rate']?.toString() ?? '',
    lprAdjust: json['lpr_adjust']?.toString() ?? '',
    nextRepricingDate: json['next_repricing_date']?.toString() ?? '',
    repaymentMethod: json['repayment_method']?.toString() ?? '',
    repaymentCard: json['repayment_card']?.toString() ?? '',
    notifyPhone: json['notify_phone']?.toString() ?? '',
  );
}

/// 养老金账户
class HeldPension {
  final String id;
  final String accountType;
  final double totalAmount;
  final double idleAmount;
  final double annualDepositLimit;
  final double remainingDepositQuota;
  final double depositedAmount;
  final String taxBenefit;
  final String autoDepositStatus;
  final String recommendedPlan;
  final String referenceYield;
  final String yieldCalculation;
  final String remark;

  HeldPension({
    required this.id,
    required this.accountType,
    required this.totalAmount,
    required this.idleAmount,
    required this.annualDepositLimit,
    required this.remainingDepositQuota,
    required this.depositedAmount,
    required this.taxBenefit,
    required this.autoDepositStatus,
    required this.recommendedPlan,
    required this.referenceYield,
    required this.yieldCalculation,
    required this.remark,
  });

  factory HeldPension.fromJson(Map<String, dynamic> json) => HeldPension(
    id: json['id']?.toString() ?? '',
    accountType: json['account_type']?.toString() ?? '',
    totalAmount: _parseDouble(json['total_amount']),
    idleAmount: _parseDouble(json['idle_amount']),
    annualDepositLimit: _parseDouble(json['annual_deposit_limit']),
    remainingDepositQuota: _parseDouble(json['remaining_deposit_quota']),
    depositedAmount: _parseDouble(json['deposited_amount']),
    taxBenefit: json['tax_benefit']?.toString() ?? '',
    autoDepositStatus: json['auto_deposit_status']?.toString() ?? '',
    recommendedPlan: json['recommended_plan']?.toString() ?? '',
    referenceYield: json['reference_yield']?.toString() ?? '',
    yieldCalculation: json['yield_calculation']?.toString() ?? '',
    remark: json['remark']?.toString() ?? '',
  );
}

/// 持有产品汇总
class HeldProductsSummary {
  final int wealthCount;
  final int loanCount;
  final int pensionCount;
  final double totalWealthAmount;
  final double totalLoanAmount;
  final double totalPensionAmount;

  HeldProductsSummary({
    required this.wealthCount,
    required this.loanCount,
    required this.pensionCount,
    required this.totalWealthAmount,
    required this.totalLoanAmount,
    required this.totalPensionAmount,
  });

  factory HeldProductsSummary.fromJson(Map<String, dynamic> json) {
    final wealth = json['wealth'] ?? {};
    final loans = json['loans'] ?? {};
    final pensions = json['pensions'] ?? {};
    return HeldProductsSummary(
      wealthCount: (wealth['count'] ?? 0) as int,
      loanCount: (loans['count'] ?? 0) as int,
      pensionCount: (pensions['count'] ?? 0) as int,
      totalWealthAmount: _parseDouble(wealth['total_holding_amount']),
      totalLoanAmount: _parseDouble(loans['total_remaining_principal']),
      totalPensionAmount: _parseDouble(pensions['total_amount']),
    );
  }
}

double _parseDouble(dynamic value) {
  if (value == null) return 0.0;
  if (value is double) return value;
  if (value is int) return value.toDouble();
  if (value is String) {
    final cleaned = value.replaceAll('%', '').replaceAll(',', '');
    return double.tryParse(cleaned) ?? 0.0;
  }
  return 0.0;
}
